compiler/src/dotty/tools/dotc/typer/Applications.scala

```diff
@@ -2133,8 +2133,45 @@ trait Applications extends Compatibility {
     }
   }
 
-  def isApplicableExtensionMethod(ref: TermRef, receiver: Type)(using Context) =
-    ref.symbol.is(ExtensionMethod)
-    && !receiver.isBottomType
-    && isApplicableMethodRef(ref, receiver :: Nil, WildcardType)
+  /** Assuming methodRef is a reference to an extension method defined e.g. as
+   *
+   *  extension [T1, T2](using A)(using B, C)(receiver: R)(using D)
+   *    def foo[T3](using E)(f: F): G = ???
+   *
+   *  return the tree representing methodRef partially applied to the receiver and all the implicit parameters preceding it (A, B, C)
+   *  with the type parameters of the extension (T1, T2) inferred.
+   *  A failure is returned if the implicit search fails for any of the leading implicit parameters or if the receiver has a wrong type
+   *  (note that in general the type of the receiver might depend on the exact types of the found instances of the proceding implicits).
+   *  No implicit search is tried for implicits following the receiver or for parameters of the def (D, E).
+   */
+  private def applyWithoutPostreceiverImplicits(methodRef: TermRef, receiver: Tree)(using Context): scala.util.Try[Tree] =
+      // Drop all parameters sections of an extension method following the receiver; the return type after truncation is not important
+    def truncateExtension(tp: Type)(using Context): Type = tp match
+      case poly: PolyType => poly.newLikeThis(poly.paramNames, poly.paramInfos, truncateExtension(poly.resType))
+      case meth: MethodType if meth.isContextualMethod => meth.newLikeThis(meth.paramNames, meth.paramInfos, truncateExtension(meth.resType))
+      case meth: MethodType => meth.newLikeThis(meth.paramNames, meth.paramInfos, defn.AnyType)
+
+    val truncatedSym = methodRef.symbol.asTerm.copy(owner = defn.RootPackage, name = Names.termName(""), info = truncateExtension(methodRef.info))
+    val truncatedRef = ref(truncatedSym).withSpan(receiver.span)
+    val newCtx = ctx.fresh.setNewScope.setReporter(new reporting.ThrowingReporter(ctx.reporter))
+    scala.util.Try {
+      inContext(newCtx) {
+        ctx.enter(truncatedSym)
+        ctx.typer.extMethodApply(truncatedRef, receiver, WildcardType)
+      }
+    }.filter(tree => tree.tpe.exists && !tree.tpe.isError)
+
+  def tryApplyingReceiver(methodRef: TermRef, receiver: Tree)(using Context): Option[Tree] =
+    def replaceCallee(inTree: Tree, replacement: Tree)(using Context): Tree = inTree match
+      case Apply(fun, args) => Apply(replaceCallee(fun, replacement), args)
+      case TypeApply(fun, args) => TypeApply(replaceCallee(fun, replacement), args)
+      case _: Ident => replacement
+
+    applyWithoutPostreceiverImplicits(methodRef, receiver)
+      .toOption
+      .map(tree => replaceCallee(tree, ref(methodRef)))
```

Comment:
I'd suggest merge `tryApplyingReceiver` and `applyWithoutPostreceiverImplicits`, maybe name it `tryApplyExtensionMethod`.

Reply:
I kept them separate for performance as `applyWithoutPostreceiverImplicits` is also reused when searching for potential import suggestions and there we are only interested in whether it was a success, not in the transformed result. Also IMHO with `tryApplyExtensionMethod` it would not be less clear that the arguments following the receiver are not passed to the method

Reply:
If we then move `.toOption` to the body of `tryApplyingExtensionMethod1` then the performance optimization is lost. If we don't care about it (don't we?) then keeping the method split doesn't make sense

Reply:
I don't think this micro-optimization matters, having a more clear return type is more important. The current result type `Try[Tree]` provides more information than necessary, and can be safely replaced with `Option[Tree]`

Reply:
BTW, the code can use `try/catch` to return `Option[Tree]` without the overhead of `.toOption`.